### PR TITLE
Limit CI timeout for base and model testsuites

### LINKED ISSUE #15118: Build Timeouts on integration tests
### Describe the bug

Builds for PRs fail with a timeout after 6 hours. 

See https://github.com/keycloak/keycloak/actions/runs/3284539014/jobs/5413063587 for an example.

### Version

nightly

### Expected behavior

Builds should complete within ~60 min, 90 minutes max.

### Actual behavior

Builds time out after 6 hours

### How to Reproduce?

This is happening only sometimes, needs to be investigated.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by hmlnarik**
Fixes: #15118

### CODE REVIEW

**Review by ahus1 — DISMISSED** [review]
Thank you for this PR to limit maximum run time.

**Comment by ahus1** [comment]
@hmlnarik - all checks have now passed.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -58,6 +58,7 @@ jobs:
     name: Unit Tests
     runs-on: ubuntu-latest
     needs: build
+    timeout-minutes: 90
```
Comment (stianst) [inline .github/workflows/ci.yml:61]:
```suggestion
    timeout-minutes: 30
```

Last run was 9 min, maybe we could lower this to 30?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -104,6 +105,7 @@ jobs:
     name: Crypto Tests
     runs-on: ubuntu-latest
     needs: build
+    timeout-minutes: 90
```
Comment (stianst) [inline .github/workflows/ci.yml:108]:
```suggestion
    timeout-minutes: 30
```

Last run was <2 min, maybe we could lower this to 30?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -148,6 +150,7 @@ jobs:
     name: Model Tests
     runs-on: ubuntu-latest
     needs: build
+    timeout-minutes: 90
```
Comment (stianst) [inline .github/workflows/ci.yml:153]: Am I correct in thinking that model tests profiles depend on what was changed? Or, are all profiles always executed?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -278,6 +282,7 @@ jobs:
     name: Base testsuite (fips)
     needs: build
     runs-on: ubuntu-latest
+    timeout-minutes: 90
```
Comment (stianst) [inline .github/workflows/ci.yml:285]:
```suggestion
    timeout-minutes: 60
```

Last run was in 20 min, perhaps 60 is more than sufficient?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -350,6 +355,7 @@ jobs:
     name: Base testsuite (postgres)
     needs: build
     runs-on: ubuntu-latest
+    timeout-minutes: 90
```
Comment (stianst) [inline .github/workflows/ci.yml:358]: Last run had one matrix above 60 min, so 90 is good here I'd say.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -448,6 +454,7 @@ jobs:
     name: Quarkus Test Clustering
     needs: build
     runs-on: ubuntu-latest
+    timeout-minutes: 90
```
Comment (stianst) [inline .github/workflows/ci.yml:457]:
```suggestion
    timeout-minutes: 60
```

Last run was 22 min, so 60 should be more than sufficient.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -508,6 +515,7 @@ jobs:
     name: Quarkus Tests
     needs: build
     runs-on: ubuntu-latest
+    timeout-minutes: 120
```
Comment (stianst) [inline .github/workflows/ci.yml:518]: Last run was 1h 17m, so guess 90 is not sufficient, while 120 is.

**Review by stianst — COMMENTED** [review]
Suggested some lower numbers for some tests. Wondering if we could even go more aggressive than that, and only add a small margin on top of the current execution time?

I'd actually be tempted to go quite low (current time plus say 10 min?), like for example:

* Unit Tests: 20 min
* Base Postgres: 75 min
* Quarkus Test Clustering: 35 min
* Quarkus Tests: 90 min

Not something I'm pushing for, but worth discussing?

File: .github/workflows/ci.yml (hunk 3 above)
Comment (hmlnarik) [reply .github/workflows/ci.yml:153]: All profiles are always executed. There is no straightforward way to map profiles to changes yet.

**Comment by hmlnarik** [comment]
> Suggested some lower numbers for some tests. Wondering if we could even go more aggressive than that, and only add a small margin on top of the current execution time?
> 
> I'd actually be tempted to go quite low (current time plus say 10 min?), like for example:
> 
> * Unit Tests: 20 min
> * Base Postgres: 75 min
> * Quarkus Test Clustering: 35 min
> * Quarkus Tests: 90 min
> 
> Not something I'm pushing for, but worth discussing?

Having a lower cap on running time would be beneficial for GH runner utilization, I agree.

I would take thus this PR as an medium-term solution which improves the current state by pushing the limits to one third of today, and then have a follow-up issue for analysis of a 50-100 of last runs which would help in establishing the timeouts say by job median time + 10 minutes. WDYT?

**Comment by stianst** [comment]
> > Suggested some lower numbers for some tests. Wondering if we could even go more aggressive than that, and only add a small margin on top of the current execution time?
> > I'd actually be tempted to go quite low (current time plus say 10 min?), like for example:
> > 
> > * Unit Tests: 20 min
> > * Base Postgres: 75 min
> > * Quarkus Test Clustering: 35 min
> > * Quarkus Tests: 90 min
> > 
> > Not something I'm pushing for, but worth discussing?
> 
> Having a lower cap on running time would be beneficial for GH runner utilization, I agree.
> 
> I would take thus this PR as an medium-term solution which improves the current state by pushing the limits to one third of today, and then have a follow-up issue for analysis of a 50-100 of last runs which would help in establishing the timeouts say by job median time + 10 minutes. WDYT?

That's never going to happen. I'd rather just go more aggressive now, then increase slightly if it becomes a problem.

**Comment by hmlnarik** [comment]
> That's never going to happen. I'd rather just go more aggressive now, then increase slightly if it becomes a problem.

```bash
mkdir /tmp/gh
IDS="$(gh run list --workflow 'Keycloak CI' -b main --jq '.[].databaseId ' --json databaseId)"
for I in "$IDS"; do gh run view $I > /tmp/gh/$I.out; done
## Output is in format "ID;minutes;seconds;job-name;success-failure-flag"
perl -ne 'print "$+{id};" . ($3 * 60 + $4) . ";$5;$2;$1\n" if (/^(.*?) (.*) in (?:(\d+)h)?(?:(\d+)m)(?:(\d+)s)? \(ID (?<id>\d+)\)/)' * | sort -t \; -k4 > times.csv
```

Then import into a spreadsheet calculator and create a pivot table out of it.

Good estimates still, they mostly match. 👍🏻 

I have updated the timeouts according to successful runs.

**Review by stianst — APPROVED** [review]

**Comment by stianst** [comment]
Thanks, happy with this, and nice to see there's an easy way to get the run times.

**Comment by stianst** [comment]
Related to this I suggest we do a lot more scheduled runs (at least temporarily) so we can get some more info on what is unstable:
https://github.com/keycloak/keycloak/pull/15273

**Pull request merged**